url/url.cpp

```diff
@@ -51,4 +101,1381 @@ std::string blob_url_create(Origin const &origin) {
+    password_token_seen_ = false;
+
+    // Initialize BaseParser with our modified input
+    reset(input);
+
+    while (true) {
+        switch (state_) {
+            case ParserState::SchemeStart:
+                state_scheme_start();
+                break;
+            case ParserState::Scheme:
+                state_scheme();
+                break;
+            case ParserState::NoScheme:
+                state_no_scheme();
+                break;
+            case ParserState::SpecialRelativeOrAuthority:
+                state_special_relative_or_authority();
+                break;
+            case ParserState::PathOrAuthority:
+                state_path_or_authority();
+                break;
+            case ParserState::Relative:
+                state_relative();
+                break;
+            case ParserState::RelativeSlash:
+                state_relative_slash();
+                break;
+            case ParserState::SpecialAuthoritySlashes:
+                state_special_authority_slashes();
+                break;
+            case ParserState::SpecialAuthorityIgnoreSlashes:
+                state_special_authority_ignore_slashes();
+                break;
+            case ParserState::Authority:
+                state_authority();
+                break;
+            case ParserState::Host:
+                state_host();
+                break;
+            case ParserState::Port:
+                state_port();
+                break;
+            case ParserState::File:
+                state_file();
+                break;
+            case ParserState::FileSlash:
+                state_file_slash();
+                break;
+            case ParserState::FileHost:
+                state_file_host();
+                break;
+            case ParserState::PathStart:
+                state_path_start();
+                break;
+            case ParserState::Path:
+                state_path();
+                break;
+            case ParserState::OpaquePath:
+                state_opaque_path();
+                break;
+            case ParserState::Query:
+                state_query();
+                break;
+            case ParserState::Fragment:
+                state_fragment();
+                break;
+            case ParserState::Failure:
+                return std::nullopt;
+            case ParserState::Terminate:
+                // I use this state where the spec returns "nothing" (i.e, the parser is modifying a given optional URL)
+                // Instead of modifying it in-place, I modify a copy and return that instead of nothing.
+                return url_;
+            default:
+                return std::nullopt;
+        }
+
+        if (is_eof()) {
+            break;
+        }
+
+        advance(1);
+    }
+
+    return url_;
+}
+
+// https://url.spec.whatwg.org/#scheme-start-state
+void UrlParser::state_scheme_start() {
+    if (peek().has_value() && util::is_alpha(*peek())) {
```

Comment:
If I'm going to change this, I'm going to do it by exposing a `c` field from the baseparser, or something like that. I don't think this is the right way to handle it; I use `peek()` this way in about a million places.

Reply:
It's not about it being too much work, it just seems like a less elegant solution than caching the value of peek() when the pointer gets updated. "less typing" and "less duplication" would also be using `c_` everywhere in the code instead of just in the condition for an if-statement where I check the value of `peek()` multiple times, so I don't understand the objection.